pkg/apis/proto/source/v1/udsource.proto

```diff
@@ -0,0 +1,138 @@
+syntax = "proto3";
+
+option go_package = "github.com/numaproj/numaflow-go/pkg/apis/proto/source/v1";
+
+import "google/protobuf/timestamp.proto";
+import "google/protobuf/empty.proto";
+
+package source.v1;
+
+service UserDefinedSource {
+  // Read returns a stream of datum responses.
+  // The size of the returned ReadResponse is less than or equal to the num_records specified in ReadRequest.
+  rpc Read(ReadRequest) returns (stream ReadResponse);
+
+  // Ack acknowledges a list of datum offsets.
+  // It indicates that the datum stream has been processed by the source vertex.
+  rpc Ack(AckRequest) returns (AckResponse);
+
+  // Pending returns the number of pending records at the user defined source.
+  rpc Pending(google.protobuf.Empty) returns (PendingResponse);
+
+  // IsSourceReady is the heartbeat endpoint for user defined source gRPC.
+  rpc IsSourceReady(google.protobuf.Empty) returns (SourceReadyResponse);
+}
+
+/*
+ * EventTime represents the event time of the datum.
+ */
+message EventTime {
+  google.protobuf.Timestamp event_time = 1;
+}
+
+/*
+ * ReadRequest is the request for reading datum stream from user defined source.
+ */
+message ReadRequest {
+  message Request {
+    // Required field indicating the number of records to read.
+    uint64 num_records = 1;
```

Comment:
Shall we accommodate the options such as `readTimeOut`?

Reply:
That's a good point. I am thinking we can start from simple and add more options at a later time. WDYT?

Reply:
The timeout is something required for reading logic, even it's marked as optional in our code, just because there's a default value. Without it, anyone who implements a udsource will have to give a timeout, with this, I think we should put it in the protocol, and we will make sure it always has a value.